src/chttpd/src/chttpd.erl

```diff
@@ -1228,6 +1236,110 @@ maybe_trace_fdb("true") ->
 maybe_trace_fdb(_) ->
     ok.
 
+start_span(Req) ->
+    #httpd{
+        mochi_req = MochiReq,
+        begin_ts = Begin,
+        peer = Peer,
+        nonce = Nonce,
+        method = Method,
+        path_parts = PathParts
+    } = Req,
+    {OperationName, ExtraTags} = get_action(Req),
+    Tags = maps:merge(#{
+        peer => Peer,
+        'http.method' => Method,
+        nonce => Nonce,
+        'http.url' => MochiReq:get(raw_path),
+        path_parts => PathParts,
+        'span.kind' => <<"server">>,
+        component => <<"couchdb.chttpd">>
+    }, ExtraTags),
+
+    ctrace:start_span(OperationName, [
+        {tags, Tags},
+        {time, Begin}
+    ] ++ maybe_root_span(MochiReq)).
+
+maybe_root_span(MochiReq) ->
+    case get_trace_headers(MochiReq) of
+        [undefined, _, _] ->
+            [];
+        [TraceId, SpanId, ParentSpanId] ->
+            Span = ctrace:external_span(TraceId, SpanId, ParentSpanId),
+            [{root, Span}]
+    end.
+
+parse_trace_id(undefined) ->
+    undefined;
+parse_trace_id(Hex) ->
+    to_int(Hex, 32).
+
+parse_span_id(undefined) ->
+    undefined;
+parse_span_id(Hex) ->
+    to_int(Hex, 16).
+
+to_int(Hex, N) when length(Hex) =:= N ->
+    try
+        list_to_integer(Hex, 16)
+    catch error:badarg ->
+        undefined
+    end.
+
+get_trace_headers(MochiReq) ->
+    case MochiReq:get_header_value("b3") of
+        undefined ->
+            [
+                parse_trace_id(MochiReq:get_header_value("X-B3-TraceId")),
+                parse_span_id(MochiReq:get_header_value("X-B3-SpanId")),
+                parse_span_id(MochiReq:get_header_value("X-B3-ParentSpanId"))
+            ];
+        Value ->
+            case binary:split(Value, <<"-">>, [global]) of
+                [TraceIdStr, SpanIdStr, _SampledStr, ParentSpanIdStr] ->
+                    [
+                        parse_trace_id(TraceIdStr),
+                        parse_span_id(SpanIdStr),
+                        parse_span_id(ParentSpanIdStr)
+                    ];
+                _ ->
+                    [undefined, undefined, undefined]
+             end
+    end.
+
+get_action(#httpd{} = Req) ->
+    try
+        chttpd_handlers:handler_info(Req)
+    catch Tag:Error ->
+        couch_log:error("Cannot set tracing action ~p:~p", [Tag, Error]),
+        {undefind, #{}}
+    end.
+
+finish_span(Fun) ->
+    Fun(),
+    ctrace:finish_span().
+
+span_ok(#httpd_resp{code = Code}) ->
+    fun() ->
```

Comment:
We do this optimization in [`ctrace:finish_span`](https://github.com/cloudant/couchdb/blob/opentracing-pd/src/ctrace/src/ctrace.erl#L159)
```
finish_span(Options0) ->
    case get(?ACTIONS_KEY) of
        undefined ->
            ok;
        Actions ->
            Options = add_time(Options0),
            lists:takewhile(fun(Action) ->
                Action(passage_pd:current_span(), Options)
            end, Actions)
    end.
```

The `?ACTIONS_KEY` is on the process dictionary only if there was a successful match on [`ctrace:start_span`](https://github.com/cloudant/couchdb/blob/opentracing-pd/src/ctrace/src/ctrace.erl#L139).

Are you proposing to pass list of extra tags to finish_span instead of closure? This would work. The one issue with this would be how to emit ctrace:log if we would want to. Since we don't use the logs at the moment I can make it possible to pass extra tags.

Reply:
I was just saying that the finish_span isn't executing any logic when invoking the closure. So the whole `finish_span(span_ok(...))` is basically `(fun() -> stuff)()` which begs the question why not just execute `stuff` without wrapping in a closure.

Alternatively, if `ctrace:finish_span` accepted a closure I could see it in that case so that the work in the closure doesn't happen until after you check `?ACTIONS_KEY`.